src/solvers/lowering/byte_operators.cpp

```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+        result = with_exprt(
+          result,
+          where,
+          typecast_exprt::conditional_cast(conditional_element, subtype));
+      }
+      else
+      {
+        result.add_to_operands(
+          where,
+          typecast_exprt::conditional_cast(conditional_element, subtype));
+      }
+    }
+  }
 
-  INVARIANT_WITH_DIAGNOSTICS(
-    element_size_opt.has_value(),
-    "size of type in bytes must be known",
-    irep_pretty_diagnosticst(src));
+  return simplify_expr(result, ns);
+}
 
-  const mp_integer &element_size = *element_size_opt;
+static exprt lower_byte_update_array_vector_non_const(
+  const byte_update_exprt &src,
+  const typet &subtype,
+  const array_exprt &value_as_byte_array,
+  const optionalt<exprt> &non_const_update_bound,
+  const namespacet &ns)
+{
+  const irep_idt extract_opcode = src.id() == ID_byte_update_little_endian
+                                    ? ID_byte_extract_little_endian
+                                    : ID_byte_extract_big_endian;
+
+  const exprt subtype_size = size_of_expr(subtype, ns);
+
+  // compute the index of the first element of the array/vector that may be
+  // updated
+  exprt first_index = div_exprt(
+    typecast_exprt::conditional_cast(src.offset(), subtype_size.type()),
+    subtype_size);
+  simplify(first_index, ns);
+
+  // compute the offset within that first element
+  const exprt update_offset = mod_exprt(
+    typecast_exprt::conditional_cast(src.offset(), subtype_size.type()),
+    subtype_size);
+
+  // compute the number of bytes (from the update value) are going to be
+  // consumed for updating the first element
+  exprt initial_bytes = minus_exprt(subtype_size, update_offset);
+  exprt update_bound;
+  if(non_const_update_bound.has_value())
+    update_bound = typecast_exprt::conditional_cast(
+      *non_const_update_bound, subtype_size.type());
+  else
+    update_bound =
+      from_integer(value_as_byte_array.operands().size(), initial_bytes.type());
+  initial_bytes = if_exprt(
+    binary_predicate_exprt(initial_bytes, ID_lt, update_bound),
+    initial_bytes,
+    update_bound);
+  simplify(initial_bytes, ns);
+
+  // encode the first update: update the original element at first_index with
+  // initial_bytes-many bytes extracted from value_as_byte_array
+  with_exprt result(
+    src.op(),
+    first_index,
+    lower_byte_operators(
+      byte_update_exprt(
+        src.id(),
+        index_exprt(src.op(), first_index),
+        update_offset,
+        byte_extract_exprt(
+          extract_opcode,
+          value_as_byte_array,
+          from_integer(0, src.offset().type()),
+          array_typet(unsignedbv_typet(8), initial_bytes))),
+      ns));
+
+  // we will update one array/vector element at a time - compute the number of
+  // update values that will be consumed in each step; if we cannot determine a
+  // constant value at this time we assume it's just one byte
+  std::size_t step_size = 1;
+  if(subtype_size.is_constant())
```

Comment:
I'm confused what you intend to happen / how this is supposed to work when this is non-constant; suggest commenting to that effect.

Reply:
Code amended to use a symbolic value and comment clarified.